File: 202-happy-number/happy-number.java
class Solution {
    public boolean isHappy(int n) 
    {  
        //Mehnaz Java Solution
        //HashSet and dunction method
        //Time complexity = O(log n)
        //Space Complexity = O(log n) - usage of HashSet

        HashSet<Integer> seen = new HashSet<>();
        while( n != 1)
        {
            if(seen.contains(n))
            {
                return false; // we have detected infinite loop
            }
            else
            {
                seen.add(n);
                n = getSum(n);
            }
        }
      return true;
    }
    public static int getSum(int num)
    {
        int sum = 0;
        while(num > 0)
        {
            int digit = num % 10;
            sum += digit * digit;
            num /= 10;
        }
        return sum;
    }
}

File: 202-happy-number/happy-number.py
class Solution(object):
    def isHappy(self, n):
        seen = set()
        while n != 1:
            if n in seen:
                return False
            seen.add(n)
            n = self.happyNum(n)
        return n == 1
    def happyNum(self, n):
        total = 0
        while n != 0:
            digit = n % 10
            total += digit * digit
            n //= 10
        return total

    



        